### ISSUE TITLE
Calculation of Anomaly Effect (∆Shigh)

### ISSUE THREAD

**Issue by blackboxo** [squareRoot3/Rethinking-Anomaly-Detection#8]
I want to reproduce the calculation of Anomaly Effect (∆Shigh)  since you haven't provided the code of this part. But I have got different results from my calculation. I am wondering where the difference is. Could you please check my code whether it has any problems? Thanks!

My result of mean and lowest value of Amazon is -0.43%(-8.74%), the result provided in the paper is -0.59% (-4.19%).
```
def laplacian(A):
    n, m = A.shape
    diags = A.sum(axis=1).flatten()
    D = scipy.sparse.spdiags(diags, [0], m, n, format="csr")
    L = D - A
    with scipy.errstate(divide="ignore"):
        diags_sqrt = 1.0 / np.sqrt(diags)
    diags_sqrt[np.isinf(diags_sqrt)] = 0
    DH = scipy.sparse.spdiags(diags_sqrt, [0], m, n, format="csr")
    return DH @ (L @ DH)

# Get effect of original graph
graph = Dataset('amazon').graph
label = graph.ndata['label'].numpy()
L = laplacian(graph.adj(scipy_fmt='csr'))

score = []
feature = graph.ndata['feature'].float()

for i in range(0,25):
    x = feature[:,i].numpy()
    xt = x.reshape(-1,1)
    x = scipy.sparse.csr_matrix(x)
    xt = scipy.sparse.csr_matrix(xt)
    result = (x.dot(L)).dot(xt) / x.dot(xt)
    score.append(result.A[0][0])

# Get effect of excluding anomalies
all_normal = list((1 - label).nonzero()[0])
normal_graph = dgl.node_subgraph(graph, all_normal)
normal_L = laplacian(normal_graph.adj(scipy_fmt='csr'))

normal_score = []
feature = normal_graph.ndata['feature'].float()

for i in range(0,25):
    x = feature[:,i].numpy()
    xt = x.reshape(-1,1)
    x = scipy.sparse.csr_matrix(x)
    xt = scipy.sparse.csr_matrix(xt)
    result = (x.dot(normal_L)).dot(xt) / x.dot(xt)
    normal_score.append(result.A[0][0])

mean = 0
low = 100

for i in range(0,25):
    delta = (normal_score[i] - score[i]) / score[i]
    mean += delta
    low = min(low, delta)
    
print(mean/25)
print(low)
```

**Comment by squareRoot3** [comment]
Hi! You can try this code. I'm too busy to check the difference with your code. Maybe due to feature normalization.

```python

import dgl
import random
import numpy as np
from dataset import *
import scipy
from matplotlib import pyplot as plt
seed = 0
torch.manual_seed(seed)
np.random.seed(seed)
random.seed(seed)


def laplacian(A):
    n, m = A.shape
    diags = A.sum(axis=1).flatten()
    D = scipy.sparse.spdiags(diags, [0], m, n, format="csr")
    L = D - A
    with scipy.errstate(divide="ignore"):
        diags_sqrt = 1.0 / np.sqrt(diags)
    diags_sqrt[np.isinf(diags_sqrt)] = 0
    DH = scipy.sparse.spdiags(diags_sqrt, [0], m, n, format="csr")
    return DH @ (L @ DH)


data_name = 'amazon'

graph = Dataset(data_name).graph
feature = graph.ndata['feature'].numpy()
label = graph.ndata['label'].numpy()
adjmat = graph.adj(scipy_fmt='csr')
L = laplacian(adjmat)

feature = (feature - feature.min(0))/feature.max(0)
graph.ndata['feature'] = torch.tensor(feature)

all_anomaly = list(label.nonzero()[0])
all_normal = list((1-label).nonzero()[0])
normal_graph = dgl.node_subgraph(graph, all_normal)
normal_feature = normal_graph.ndata['feature'].numpy()
normal_L = laplacian(normal_graph.adj(scipy_fmt='csr'))
# anomaly_feature_ave = feature[all_anomaly].mean(0)
# normal_feature_ave = feature[all_normal].mean(0)

best_diff, best_id = -1, -1
area_list = []
for i in range(feature.shape[1]):
    feat = feature[:, i]
    feat2 = normal_feature[:, i]
    print(feat2, feat2.shape, normal_L.shape)
    area1 = feat.dot(L.dot(feat))/feat.dot(feat)
    area2 = feat2.dot(normal_L.dot(feat2))/feat2.dot(feat2)
    print(area1, area2, (area1-area2)/area1, i)
    area_list.append((area1-area2)/area1)
    if best_diff < (area1-area2)/area1:
        best_diff = (area1-area2)/area1
        best_id = i
print('id: ', best_id)
print('best diff: ', best_diff)
print('average: ', np.mean(area_list))


```

**Comment by blackboxo** [comment]
Thanks for your reply. I have run you code and it provides -0.54%(-4.22%), which is still different from -0.59% (-4.19%) in the paper.

**Comment by squareRoot3** [comment]
The result is same to the paper on my computer.
<img width="260" alt="image" src="https://github.com/squareRoot3/Rethinking-Anomaly-Detection/assets/13365477/8dac80f6-f92c-40c6-af89-bc3e53344da3">

**Comment by blackboxo** [comment]
that's quite strange...

**Comment by blackboxo** [comment]
that's quite strange...
